### PR TITLE
Fixes user setting for AWS telemetry, auto converts to bool

### PR DESCRIPTION
**Pull request by gsweetwood**
## Description

A function has been added executes on activation that reads the user's setting for aws.telemtry and converts the value to a boolean (if it is not already).  This respects the previous choice and will convert to false if the value was 'Disable', all other values will be converted to true.

## Motivation and Context

The setting used to be a string, it is now a boolean.  While the old value is still respected (meaning their opt-in status did not change), an error message is displayed in the settings.json file and a user would not be able to change the setting by clicking as the checkbox for the setting becomes disabled.  This change corrects the format for that setting, which removes the error message and restores the checkbox. 

## Screenshots (if appropriate)
This was the error message in settings
error message

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/telemetry/activation.ts [hunk 1]
```diff
@@ -72,24 +75,28 @@ export async function activate(activateArguments: {
         activateArguments.extensionContext.subscriptions
     )
 }
-
-export function isTelemetryEnabled(toolkitSettings: SettingsConfiguration): boolean {
+/*
+ * Formats the AWS telemetry setting to a boolean: false if setting value was 'Disable' or false, true for everything else
+ */
+export async function formatTelemetrySettingToBool(toolkitSettings: SettingsConfiguration): Promise<void> {
```
Comment (awschristou) [inline src/shared/telemetry/activation.ts:81]: Consider a less verbose but more concise name like `sanitizeTelemetrySetting`

File: src/shared/telemetry/activation.ts [hunk 2]
```diff
@@ -72,24 +75,28 @@ export async function activate(activateArguments: {
...
+        await toolkitSettings.writeSetting<any>(AWS_TELEMETRY_KEY, false, vscode.ConfigurationTarget.Global)
+    } else {
+        await toolkitSettings.writeSetting<any>(AWS_TELEMETRY_KEY, true, vscode.ConfigurationTarget.Global)
     }
+}
 
-    // Treat anything else (unexpected values, datatypes, or undefined) as opt-in
-    return true
+export function isTelemetryEnabled(toolkitSettings: SettingsConfiguration): boolean {
+    return toolkitSettings.readSetting<any>(AWS_TELEMETRY_KEY)
```
Comment (awschristou) [inline src/shared/telemetry/activation.ts:99]: The value _should_ be sanitized, however one could still go into the file and put any value the like here. It may be worth retaining the previous function implementation to prevent calling code from having unexpected data.

File: src/test/shared/telemetry/activation.test.ts [hunk 3]
```diff
@@ -103,6 +104,7 @@ describe('isTelemetryEnabled', () => {
         it(scenario.desc, async () => {
             await settings.update('telemetry', scenario.initialSettingValue, vscode.ConfigurationTarget.Global)
 
+            await formatTelemetrySettingToBool(toolkitSettings)
```
Comment (awschristou) [inline src/test/shared/telemetry/activation.test.ts:107]: If you do revise the PR's impl of isTelemetryEnabled, please test formatTelemetrySettingToBool in isolation by verifying toolkitSettings directly

File: src/shared/telemetry/activation.ts (hunk 2 above)
Comment (gsweetwood) [reply src/shared/telemetry/activation.ts:99]: Great point, I'll restore the functionality.

Comment (ckant) [reply src/shared/telemetry/activation.ts:99]: This could also be true of any value we read from the config, since VSCode doesn't throw an exception if it violates the schema. Perhaps we should either

1. Throw an error if the value is unexpected
2. Default to turning off the telemetry if the value is unexpected

File: src/shared/telemetry/activation.ts [hunk 4]
```diff
@@ -72,24 +75,28 @@ export async function activate(activateArguments: {
...
+export async function formatTelemetrySettingToBool(toolkitSettings: SettingsConfiguration): Promise<void> {
     // Setting used to be an enum, but is now a boolean.
     // We don't have api-based strong type support, so we have to process this value manually.
     const value = toolkitSettings.readSetting<any>(AWS_TELEMETRY_KEY)
 
-    // Handle original opt-out value (setting used to be a tri-state string value)
-    if (value === LEGACY_SETTINGS_TELEMETRY_VALUE_DISABLE) {
-        return false
-    }
+    if (typeof value === 'boolean') return
```
Comment (ckant) [inline src/shared/telemetry/activation.ts:86]: Please add braces around all `if` statements even if they're a single line. It's defensive against accidentally leaving stuff out of the statement which leads to unexpected behavior.

We follow this convention everywhere except here apparently 🤔 
https://github.com/aws/aws-toolkit-vscode/blob/41365536e9e580239eeff931f6956a38eb83cd01/src/eventSchemas/commands/downloadSchemaItemCode.ts#L377

File: src/shared/telemetry/activation.ts [hunk 5]
```diff
@@ -72,24 +75,28 @@ export async function activate(activateArguments: {
...
-    // Current value is expected to be a boolean
-    if (typeof value === 'boolean') {
-        return value
+    // Original setting used to be a tri-state string value
+    // Set to false if original setting was opt-out
+    // Treat anything else (unexpected values, datatypes, or undefined) as opt-in
+    if (value === LEGACY_SETTINGS_TELEMETRY_VALUE_DISABLE) {
+        await toolkitSettings.writeSetting<any>(AWS_TELEMETRY_KEY, false, vscode.ConfigurationTarget.Global)
+    } else {
+        await toolkitSettings.writeSetting<any>(AWS_TELEMETRY_KEY, true, vscode.ConfigurationTarget.Global)
```
Comment (ckant) [inline src/shared/telemetry/activation.ts:94]: This will unintentionally opt users into telemetry if they've got a typo in their config beyond the 'Enabled' 'Disabled. 

e.g. they have 'Foo' or 123.

I would swap the ordering of these. Check for Enabled, otherwise turn off telemetry. Much safer to assume an opt out.

File: src/test/shared/telemetry/activation.test.ts (hunk 3 above)
Comment (ckant) [reply src/test/shared/telemetry/activation.test.ts:107]: Yes, please write a unit test for this.

File: src/shared/telemetry/activation.ts (hunk 2 above)
Comment (awschristou) [reply src/shared/telemetry/activation.ts:99]: I prefer 1 since the point of this method is to query, not to modify

File: src/shared/telemetry/activation.ts (hunk 5 above)
Comment (awschristou) [reply src/shared/telemetry/activation.ts:94]: That is not consistent with the rest of the toolkit's handling of telemetry

File: src/shared/telemetry/activation.ts (hunk 1 above)
Comment (gsweetwood) [reply src/shared/telemetry/activation.ts]: Changed to your suggestion.

File: src/test/shared/telemetry/activation.test.ts (hunk 3 above)
Comment (gsweetwood) [reply src/test/shared/telemetry/activation.test.ts]: Great suggestion, I added a new unit test and reorganized the testing suite.

File: src/shared/telemetry/activation.ts (hunk 5 above)
Comment (ckant) [reply src/shared/telemetry/activation.ts:94]: You mean auto opt in? I'd say this edge case situation is different in that there's a value there but it's just invalid. Safer to assume this was an attempt to opt out than to assume it wasn't.

File: src/shared/telemetry/activation.ts [hunk 6]
```diff
@@ -83,13 +105,18 @@ export function isTelemetryEnabled(toolkitSettings: SettingsConfiguration): bool
         return false
     }
 
-    // Current value is expected to be a boolean
-    if (typeof value === 'boolean') {
-        return value
+    //Will throw an error if type is not a boolean
+    try {
+        if (typeof value !== 'boolean') {
+            throw new TypeError(`${value} is not a boolean`)
```
Comment (ckant) [inline src/shared/telemetry/activation.ts:111]: The option where we throw was suggesting that the exception would escape uncaught.

Honestly the safest thing to do would be to only change `Enabled` and `Disabled` to `true` and `false` resp. Would leave all other unexpected values as-is as it's the least surprising to the user.

If an unexpected value is read, I would treat it as an attempt to opt out and return `false` here. This setting is special in that assuming a `true` value could lose user trust whereas assuming a `false` value cannot. Hence why I said it was the safer of the two options.

File: src/test/shared/telemetry/activation.test.ts [hunk 7]
```diff
@@ -99,12 +100,26 @@ describe('isTelemetryEnabled', () => {
...
+
+    describe('sanitizeTelemetrySetting', () => {
+        scenarios.forEach(scenario => {
+            it(scenario.desc, async () => {
+                await settings.update('telemetry', scenario.initialSettingValue, vscode.ConfigurationTarget.Global)
 
-            const isEnabled = isTelemetryEnabled(toolkitSettings)
-            assert.strictEqual(isEnabled, scenario.expectedIsEnabledValue)
+                await sanitizeTelemetrySetting(toolkitSettings)
+                const sanitiezedSetting = toolkitSettings.readSetting<any>('telemetry')
```
Comment (ckant) [inline src/test/shared/telemetry/activation.test.ts:120]: typo

File: src/shared/telemetry/activation.ts (hunk 6 above)
Comment (gsweetwood) [reply src/shared/telemetry/activation.ts:111]: I agree that we should only change the user setting if it exactly matches 'Enable' or 'Disable'.  I will go ahead and make that change.  As for the default for unexpected values, I know there has been a change in this recently.  Let me get back to you with more input.

File: src/shared/telemetry/activation.ts [hunk 8]
```diff
@@ -83,13 +105,18 @@ export function isTelemetryEnabled(toolkitSettings: SettingsConfiguration): bool
...
+        } else {
+            return value
+        }
+    } catch (error) {
```
Comment (awschristou) [inline src/shared/telemetry/activation.ts:115]: the try/throw/catch ceremony here is overkill. When the type is bool, you can return the value, else log and return true.
